File: pufferlib/ocean/pool/pool.c
#include <stdlib.h>
#include "pool.h"

int main() {
    Pool env = {
        .table_width = 2.84f,
        .table_height = 1.42f,
        .ball_radius = 0.03f,
        .pocket_radius = 0.06f,
        .friction = 0.992f,
        .restitution = 0.96f,
        .impulse = 0.12f,
        .min_power = 0.35f,
        .reward_step = 0.0f,
        .reward_shot = -0.001f,
        .reward_legal_pot = 0.03f,
        .reward_illegal_pot = -0.015f,
        .reward_foul = -0.05f,
        .reward_win = 1.0f,
        .fast_forward = 0,
        .max_physics_steps = 96,
        .max_steps = 200,
    };

    env.observations = (float*)calloc(NUM_PLAYERS * OBS_SIZE, sizeof(float));
    env.actions = (int*)calloc(NUM_PLAYERS * 2, sizeof(int));
    env.rewards = (float*)calloc(NUM_PLAYERS, sizeof(float));
    env.terminals = (unsigned char*)calloc(NUM_PLAYERS, sizeof(unsigned char));
    env.truncations = (unsigned char*)calloc(NUM_PLAYERS, sizeof(unsigned char));

    init(&env);
    c_reset(&env);

    while (!WindowShouldClose()) {
        env.actions[0] = 0;
        env.actions[1] = 4;
        env.actions[2] = 0;
        env.actions[3] = 4;

        if (env.current_player == 0) {
            if (IsKeyDown(KEY_LEFT_SHIFT)) {
                if (IsKeyPressed(KEY_RIGHT)) env.actions[0] = 0;
                if (IsKeyPressed(KEY_DOWN)) env.actions[0] = 4;
                if (IsKeyPressed(KEY_LEFT)) env.actions[0] = 8;
                if (IsKeyPressed(KEY_UP)) env.actions[0] = 12;
                if (IsKeyPressed(KEY_ONE)) env.actions[1] = 0;
                if (IsKeyPressed(KEY_TWO)) env.actions[1] = 1;
                if (IsKeyPressed(KEY_THREE)) env.actions[1] = 2;
                if (IsKeyPressed(KEY_FOUR)) env.actions[1] = 3;
                if (IsKeyPressed(KEY_FIVE)) env.actions[1] = 4;
            }
        } else {
            if ((rand() % 18) == 0) {
                env.actions[2] = rand() % 16;
                env.actions[3] = rand() % 5;
            }
        }

        c_step(&env);
        c_render(&env);
    }

    free(env.observations);
    free(env.actions);
    free(env.rewards);
    free(env.terminals);
    free(env.truncations);
    c_close(&env);

    return 0;
}

File: pufferlib/ocean/pool/pool.py
import gymnasium
import numpy as np

import pufferlib
from pufferlib.ocean.pool import binding


class Pool(pufferlib.PufferEnv):
    def __init__(
        self,
        num_envs=256,
        width=2.84,
        height=1.42,
        ball_radius=0.03,
        pocket_radius=0.06,
        friction=0.992,
        restitution=0.96,
        impulse=0.12,
        min_power=0.35,
        reward_step=0.0,
        reward_shot=-0.001,
        reward_legal_pot=0.03,
        reward_illegal_pot=-0.015,
        reward_foul=-0.05,
        reward_win=1.0,
        fast_forward=True,
        max_physics_steps=96,
        max_steps=200,
        report_interval=128,
        render_mode='human',
        buf=None,
        seed=0,
    ):
        self.render_mode = render_mode
        self.num_tables = num_envs
        self.num_agents = num_envs * 2
        self.report_interval = report_interval
        self.tick = 0

        self.single_observation_space = gymnasium.spaces.Box(
            low=-np.inf,
            high=np.inf,
            shape=(89,),
            dtype=np.float32,
        )
        # Action[0]: shot direction in [0..15] around 360 degrees
        # Action[1]: power bin in [0..4]
        self.single_action_space = gymnasium.spaces.MultiDiscrete(
            [16, 5], dtype=np.int32
        )

        super().__init__(buf)

        c_envs = []
        for i in range(num_envs):
            start = 2 * i
            end = start + 2
            env_id = binding.env_init(
                self.observations[start:end],
                self.actions[start:end],
                self.rewards[start:end],
                self.terminals[start:end],
                self.truncations[start:end],
                i + seed * num_envs,
                width=width,
                height=height,
                ball_radius=ball_radius,
                pocket_radius=pocket_radius,
                friction=friction,
                restitution=restitution,
                impulse=impulse,
                min_power=min_power,
                reward_step=reward_step,
                reward_shot=reward_shot,
                reward_legal_pot=reward_legal_pot,
                reward_illegal_pot=reward_illegal_pot,
                reward_foul=reward_foul,
                reward_win=reward_win,
                fast_forward=fast_forward,
                max_physics_steps=max_physics_steps,
                max_steps=max_steps,
            )
            c_envs.append(env_id)

        self.c_envs = binding.vectorize(*c_envs)

    def reset(self, seed=None):
        self.tick = 0
        binding.vec_reset(self.c_envs, 0 if seed is None else seed)
        return self.observations, []

    def step(self, actions):
        self.actions[:] = actions
        self.tick += 1
        binding.vec_step(self.c_envs)

        info = []
        if self.tick % self.report_interval == 0:
            info.append(binding.vec_log(self.c_envs))

        return (
            self.observations,
            self.rewards,
            self.terminals,
            self.truncations,
            info,
        )

    def render(self):
        binding.vec_render(self.c_envs, 0)

    def close(self):
        binding.vec_close(self.c_envs)


def test_performance(timeout=10, atn_cache=1024):
    num_envs = 1024
    env = Pool(num_envs=num_envs)
    env.reset()

    tick = 0
    actions = np.empty((atn_cache, env.num_agents, 2), dtype=np.int32)
    actions[..., 0] = np.random.randint(0, 16, (atn_cache, env.num_agents), dtype=np.int32)
    actions[..., 1] = np.random.randint(0, 5, (atn_cache, env.num_agents), dtype=np.int32)

    import time

    start = time.time()
    while time.time() - start < timeout:
        env.step(actions[tick % atn_cache])
        tick += 1

    sps = env.num_agents * tick / (time.time() - start)
    print(f'SPS: {sps:,}')


if __name__ == '__main__':
    test_performance()
